text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+
+ 2. Define a standardized pretty-printing format for things that end up as symbols, and then encode that via Punycode in order to meet the character set restrictions. This would be rather simple. Symbol names would remain somewhat human-readable (but not very, because all separators would be stripped out). But without some kind of additional compression, symbol names would become rather long.
+
+ 3. Use the scheme from the previous bullet point but apply the compression scheme described above. We could do this but it wouldn't really be less complex than the Itanium inspired scheme proposed above.
+
+ 4. Define a standardized pretty-printing format for things that end up as symbols, compress with zstd (specially trained for Rust symbols) and encode the result as base63. This is rather simple but loses all human-readability. It's unclear how well this would compress. It would pull the zstd specification into the mangling scheme specification, as well as the pre-trained dictionary.
+
+The Itanium mangling (and by extension the scheme proposed here) could be considered somewhat arcane. But it is well-known from C++ and provides a good trade-off between readability, complexity, and length of generated symbols.
+
+# Prior art
+[prior-art]: #prior-art
+
+The mangling scheme described here is an adaptation of the [Itanium C++ ABI][itanium-mangling] scheme,
+which is the scheme used by the GCC toolchain (and clang when it's not compiling for MSVC).
+
+One notable improvement the proposed scheme makes upon Itanium mangling is explicit handling of
+unicode identifiers. The idea of using [Punycode][punycode] for this is taken from the
+[Swift][swift-gh] programming language's [mangling scheme][swift-mangling] (which is also based on
+Itanium mangling).
+
+
+[punycode]: https://tools.ietf.org/html/rfc3492
+[itanium-mangling]: http://refspecs.linuxbase.org/cxxabi-1.86.html#mangling
+[swift-gh]: https://github.com/apple/swift
+[swift-mangling]: https://github.com/apple/swift/blob/master/docs/ABI/Mangling.rst#identifiers
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+### Punycode vs UTF-8
+During the pre-RFC phase, it has been suggested that Unicode identifiers should be encoded as UTF-8 instead of Punycode on platforms that allow it. GCC, Clang, and MSVC seem to do this. The author of the RFC has a hard time making up their mind about this issue. Here are some interesting points that might influence the final decision:
+
+- Using UTF-8 instead of Punycode would make mangled strings containing non-ASCII identifiers a bit more human-readable. For demangled strings, there would be no difference.
```

Comment:
This is moot if compression (of any kind) is applied as well.

Reply:
You mean because compressed names are unreadable in any case?

Reply:
Right.